url147ce07ebe_JoshuaKissoon_Kademlia_tgz-pJ8-kademlia_simulations_RoutingTableStateTestingJ8/src/kademlia/KadServer.java

```diff
@@ -0,0 +1,357 @@
+            }
+        }
+
+        /* Send the message */
+        sendMessage(to, msg, comm);
+
+        return comm;
+    }
+
+    /**
+     * Method called to reply to a message received
+     *
+     * @param to   The Node to send the reply to
+     * @param msg  The reply message
+     * @param comm The communication ID - the one received
+     *
+     * @throws java.io.IOException
+     */
+    public synchronized void reply(Node to, Message msg, int comm) throws IOException
+    {
+        if (!isRunning)
+        {
+            throw new IllegalStateException("Kad Server is not running.");
+        }
+        sendMessage(to, msg, comm);
+    }
+
+    /**
+     * Internal sendMessage method called by the public sendMessage method after a communicationId is generated
+     */
+    private void sendMessage(Node to, Message msg, int comm) throws IOException
+    {
+        /* Use a try-with resource to auto-close streams after usage */
+        try (ByteArrayOutputStream bout = new ByteArrayOutputStream(); DataOutputStream dout = new DataOutputStream(bout);)
+        {
+            /* Setup the message for transmission */
+            dout.writeInt(comm);
+            dout.writeByte(msg.code());
+            msg.toStream(dout);
+            dout.close();
+
+            byte[] data = bout.toByteArray();
+
+            if (data.length > DATAGRAM_BUFFER_SIZE)
+            {
+                throw new IOException("Message is too big");
+            }
+
+            /* Everything is good, now create the packet and send it */
+            DatagramPacket pkt = new DatagramPacket(data, 0, data.length);
+            pkt.setSocketAddress(to.getSocketAddress());
+            socket.send(pkt);
+
+            /* Lets inform the statistician that we've sent some data */
+            this.statistician.sentData(data.length);
+        }
+    }
+
+    /**
+     * Listen for incoming messages in a separate thread
+     */
+    private void listen()
+    {
+        try
+        {
+            while (isRunning)
+            {
+                try
+                {
+                    /* Wait for a packet */
+                    byte[] buffer = new byte[DATAGRAM_BUFFER_SIZE];
+                    DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
+                    socket.receive(packet);
+
+                    /* Lets inform the statistician that we've received some data */
+                    this.statistician.receivedData(packet.getLength());
+
+                    if (this.config.isTesting())
+                    {
+                        /**
+                         * Simulating network latency
+                         * We pause for 1 millisecond/100 bytes
+                         */
+                        int pause = packet.getLength() / 100;
+                        try
+                        {
+                            Thread.sleep(pause);
+                        }
+                        catch (InterruptedException ex)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)